File: ios_app/iphone/ui/Calendar/CalendarView.swift
import SwiftUI
import shared

struct CalendarView: View {
    
    var body: some View {
        VmView({
            CalendarVm()
        }) { vm, state in
            CalendarViewInner(
                vm: vm,
                state: state
            )
        }
    }
}

private struct CalendarViewInner: View {
    
    let vm: CalendarVm
    let state: CalendarVm.State
    
    ///
    
    @State private var selectedDay: Int32? = nil
    
    var body: some View {
        
        VStack {
            
            VStack {
                
                HStack {
                    
                    ForEach(state.weekTitles, id: \.self) { weekTitle in
                        HStack {
                            Spacer()
                            Text(weekTitle.title)
                                .font(.system(size: 11))
                                .foregroundColor(weekTitle.isBusiness ? .primary: .secondary)
                            Spacer()
                        }
                    }
                }
                .padding(.top, 8)
                
                Divider()
                    .padding(.top, 2.0)
            }
            
            ScrollView(showsIndicators: false) {
                
                LazyVStack(spacing: 0) {
                    
                    ForEach(state.months, id: \.self) { month in
                        
                        HStack {
                            
                            ForEach(0..<month.emptyStartDaysCount, id: \.self) { _ in
                                ZStack {
                                }
                                .frame(minWidth: 0, maxWidth: .infinity)
                            }
                            
                            Text(month.title)
                                .frame(minWidth: 0, maxWidth: .infinity)
                                .foregroundColor(c.white)
                                .padding(.top, 24)
                                .padding(.bottom, 8)
                                .font(.system(size: 17, weight: .bold))
                            
                            ForEach(0..<month.emptyEndDaysCount, id: \.self) { _ in
                                ZStack {
                                }
                                .frame(minWidth: 0, maxWidth: .infinity)
                            }
                        }
                        
                        let weeks = month.weeks as! [[CalendarVm.MonthDay?]]
                        
                        ForEach(weeks, id: \.self) { week in
                            
                            HStack {
                                
                                ForEach(week, id: \.self) { day in
                                    
                                    if let day = day {

                                        let bgColor: Color = {
                                            if day.unixDay == selectedDay {
                                                return .blue
                                            } else if day.isToday {
                                                return .purple
                                            }
                                            return .clear
                                        }()
                                        
                                        Button(
                                            action: {
                                                selectedDay = day.unixDay
                                            },
                                            label: {
                                                
                                                VStack {
                                                    
                                                    Divider()
                                                    
                                                    Text(day.title)
                                                        .foregroundColor(day.isBusiness ? .white : .secondary)
                                                        .padding(.top, 6)
                                                    
                                                    ForEach(day.previews, id: \.self) { preview in
                                                        
                                                        Text(preview)
                                                            .padding(.horizontal, 1)
                                                            .foregroundColor(.secondary)
                                                            .font(.system(size: 11, weight: .light))
                                                            .lineLimit(1)
                                                    }
                                                }
                                                .padding(.bottom, 2)
                                                .frame(minWidth: 0, maxWidth: .infinity)
                                            }
                                        )
                                        .background(bgColor)
                                    } else {
                                        ZStack {
                                        }
                                        .frame(minWidth: 0, maxWidth: .infinity)
                                    }
                                }
                            }
                            
                            let selectedDayLocal = week.first { item in
                                if let item = item, let selectedDay {
                                    return item.unixDay == selectedDay
                                }
                                return false
                            }
                            
                            let isDaySelected: Bool = selectedDayLocal != nil
                            if isDaySelected {
                                EventsCalendarDayView(unixDay: selectedDayLocal!!.unixDay.toInt())
                                // Force update on selectedDay changes
                                    .id("EventsCalendarDayView_\(selectedDayLocal!!.unixDay)")
                            }
                        }
                    }
                }
            }
        }
        .padding(.horizontal, H_PADDING)
        .padding(.top, 6)
    }
}

File: android_app/src/main/java/me/timeto/app/ui/calendar/CalendarView.kt
package me.timeto.app.ui.calendar

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.expandVertically
import androidx.compose.animation.shrinkVertically
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.ui.HStack
import me.timeto.app.ui.H_PADDING
import me.timeto.app.ui.VStack
import me.timeto.app.ui.c
import me.timeto.app.ui.rememberVm
import me.timeto.app.ui.Divider
import me.timeto.app.ui.SpacerW1
import me.timeto.app.ui.tasks.tab.TasksTabView__PADDING_END
import me.timeto.shared.ui.calendar.CalendarVm

@Composable
fun CalendarView(
    modifier: Modifier,
) {

    val (_, state) = rememberVm {
        CalendarVm()
    }

    val selectedDay = remember {
        mutableStateOf<Int?>(null)
    }

    VStack(
        modifier = modifier
            .padding(start = H_PADDING, end = TasksTabView__PADDING_END),
    ) {

        VStack {

            HStack {

                state.weekTitles.forEach { weekTitle ->
                    Text(
                        text = weekTitle.title,
                        modifier = Modifier
                            .weight(1f),
                        color = if (weekTitle.isBusiness) c.text else c.secondaryText,
                        fontSize = 10.sp,
                        textAlign = TextAlign.Center,
                    )
                }
            }

            Divider(Modifier.padding(top = 2.dp))
        }

        LazyColumn(
            contentPadding = PaddingValues(top = 8.dp),
        ) {

            state.months.forEach { month ->

                item {

                    HStack {

                        repeat(month.emptyStartDaysCount) {
                            SpacerW1()
                        }

                        Text(
                            text = month.title,
                            modifier = Modifier
                                .weight(1f)
                                .padding(top = 16.dp, bottom = 8.dp),
                            color = c.white,
                            fontWeight = FontWeight.Bold,
                            textAlign = TextAlign.Center,
                        )

                        repeat(month.emptyEndDaysCount) {
                            SpacerW1()
                        }
                    }

                    month.weeks.forEach { week ->

                        HStack {

                            week.forEach { day ->

                                if (day == null)
                                    SpacerW1()
                                else {

                                    val bgAnimate = animateColorAsState(
                                        if (day.unixDay == selectedDay.value) c.blue
                                        else if (day.isToday) c.purple
                                        else c.transparent
                                    )

                                    VStack(
                                        modifier = Modifier
                                            .weight(1f)
                                            .background(bgAnimate.value)
                                            .clickable {
                                                selectedDay.value =
                                                    if (selectedDay.value == day.unixDay) null
                                                    else day.unixDay
                                            }
                                            .padding(bottom = 2.dp),
                                        horizontalAlignment = Alignment.CenterHorizontally,
                                    ) {

                                        Divider()

                                        Text(
                                            text = day.title,
                                            modifier = Modifier.padding(top = 6.dp),
                                            color = if (day.isBusiness) c.white else c.secondaryText,
                                        )

                                        day.previews.forEach { preview ->
                                            Text(
                                                text = preview,
                                                modifier = Modifier
                                                    .padding(horizontal = 2.dp),
                                                color = c.secondaryText,
                                                fontSize = 10.sp,
                                                lineHeight = 12.sp,
                                                fontWeight = FontWeight.Light,
                                                maxLines = 1,
                                                softWrap = false,
                                            )
                                        }
                                    }
                                }
                            }
                        }

                        val selectedDayLocal = week
                            .filterNotNull()
                            .firstOrNull { it.unixDay == selectedDay.value }

                        val isDaySelected = selectedDayLocal != null
                        AnimatedVisibility(
                            visible = isDaySelected,
                            enter = expandVertically(expandFrom = Alignment.Top),
                            exit = shrinkVertically(),
                        ) {
                            if (isDaySelected) {
                                CalendarDayView(
                                    unixDay = selectedDayLocal!!.unixDay,
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}
